gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/ComposeCompilerKotlinSupportPlugin.kt

```diff
@@ -52,6 +57,7 @@ class ComposeCompilerKotlinSupportPlugin : KotlinCompilerPluginSupportPlugin {
 
     override fun applyToCompilation(kotlinCompilation: KotlinCompilation<*>): Provider<List<SubpluginOption>> {
         val target = kotlinCompilation.target
+        composeCompilerArtifactProvider.checkTargetSupported(target)
```

Comment:
The overall code design became confusing.
Previously, `ComposeCompilerArtifactProvider` had a single responsibility of providing a version of `ComposeCompilerArtifact`.  Now `ComposeCompilerArtifactProvider` also knows how to check if a target is compatible with a Compose Compiler, but only if the default compiler is used. 
So the following configuration 
```
plugins {
  id("org.jetbrains.kotlin.multiplatform") version "1.7.20"
  id("org.jetbrains.compose") version "1.2.0"
}

kotlin {
   js(IR) {
        browser()
        binaries.executable()
    }
}

compose {
     kotlinCompilerPlugin.set("1.3.2-alpha01")
     
     web {}
}
```
will not trigger the check at all, despite that we know they are incompatible.

I suggest to separate *compiler inference* and *compiler compatibility*:
1. The inference should only try to choose a version. If inference fails, it should tell user to try updating the Compose Compiler (if a compatible version exists).
2. Then we run compiler compatibility checks, regardless of whether the version was inferred or set manually.

Reply:
> I suggest to separate compiler inference and compiler compatibility:

See my comment in the other discussion. We can't make compatibility check for an arbitrary compiler in the Gradle plugin. We only can make such check, if the compiler was chosen from our list.

As for code quality, we expand responsibility of `ComposeCompilerArtifactProvider` from `providing a version of ComposeCompilerArtifact` to `providing info about compatible compiler`. It still can be considered as a single responsibility.